File: apps/cli-dotnet/AiSkills.Cli/Core/Browse.cs
namespace AiSkills.Cli.Core;

public static class Browse
{
    public static IReadOnlyList<CatalogEntry> Filter(
        Catalog catalog,
        string? type = null,
        string? agent = null,
        string? query = null)
    {
        IEnumerable<CatalogEntry> q = catalog.Entries;

        if (!string.IsNullOrEmpty(type))
        {
            q = q.Where(e => e.Type == type);
        }

        if (!string.IsNullOrEmpty(agent))
        {
            q = q.Where(e => e.Agents.Contains(agent));
        }

        if (!string.IsNullOrEmpty(query))
        {
            var needle = query.ToLowerInvariant();
            q = q.Where(e =>
                string.Join(' ', new[] { e.Id, e.Description }.Concat(e.Tags))
                    .ToLowerInvariant()
                    .Contains(needle));
        }

        return q.ToList();
    }

    public static CatalogEntry? Find(Catalog catalog, string id) =>
        catalog.Entries.FirstOrDefault(e => e.Id == id);
}

File: apps/cli-npx/src/core/browse.ts
import type { Agent, Catalog, CatalogEntry } from '@ai-skills/catalog';

export interface BrowseFilter {
  type?: string;
  agent?: string;
  query?: string;
}

/** Filter catalog entries by type, agent, and a free-text query (id/description/tags). */
export function filterEntries(catalog: Catalog, filter: BrowseFilter = {}): CatalogEntry[] {
  return catalog.entries.filter((e) => {
    if (filter.type && e.type !== filter.type) return false;
    if (filter.agent && !e.agents.includes(filter.agent as Agent)) return false;
    if (filter.query) {
      const haystack = [e.id, e.description, ...e.tags].join(' ').toLowerCase();
      if (!haystack.includes(filter.query.toLowerCase())) return false;
    }
    return true;
  });
}

/** Find a single entry by id. */
export function findEntry(catalog: Catalog, id: string): CatalogEntry | undefined {
  return catalog.entries.find((e) => e.id === id);
}
